ninep/protocol/server.go

```diff
@@ -238,7 +238,7 @@ func (c *conn) serve() {
 
 	for !c.dead {
 		l := make([]byte, 7)
-		if n, err := c.Read(l); err != nil || n < 7 {
+		if _, err := io.ReadFull(c, l); err != nil {
```

Comment:
I'm suspicious of this fix. 7 bytes is the minimum required header for a packet. If you read less you have a problem. Could you tell me a bit more about the scenario in which you got the short reads? In general, a short read means the connection dropped. This has been a pretty reliable test for some time now, so I'd like to know more.

Reply:
I was able to reproduce with 9legacy as well. Serve 9legacy iso using ufs. In Plan 9:
```
% srv 'tcp!10.0.2.2!5640' ufs
% mount /srv/ufs /n/ufs
% cd /n/ufs/sys/src/9/pc
% while(mk clean && mk) {} 
... after 2 iteration ...
8c -FTVw -I. ../port/auth.c
../port/alarm.c:5 8c: 'dat.h' Hangup: dat.h
../port/alloc.c:3 8c: '..' Hangup: ../port/lib.h
mk: 8c -FTVw -I. ../port/alarm.c  : exit status=rc 7149: 8c 7151: error
<eof> cannot open allocb.8 - clone failed
<eof> cannot open auth.8 - clone failed
% ls
ls: .: clone failed
```
ufs debugging prints show it got 3 bytes instead of 7. Could it be that Plan 9 kernel is buffering the writes, so we need two reads to get the 7 bytes?
```
2020/09/06 14:44:51 [127.0.0.1:50650] readNetPackets: Write 11 back
2020/09/06 14:44:51 [127.0.0.1:50650] Returned 11 amt 11
2020/09/06 14:44:51 [127.0.0.1:50650] readNetPackets: got Twalk, len 18, sending to IO
2020/09/06 14:44:51 >>> Twalk fid 458, newfid 475, paths [port]
2020/09/06 14:44:51 <<< Rwalk [{128 2370785664 15993759}]
2020/09/06 14:44:51 [127.0.0.1:50650] readNetPackets: Write 22 back
2020/09/06 14:44:51 [127.0.0.1:50650] Returned 22 amt 22
2020/09/06 14:44:51 [127.0.0.1:50650] readNetPackets: short read: read 3 bytes: <nil>
```

Reply:
this needs  bit more digging for sure, b/c you can't do much with less than 7 bytes in 9p.

if you're certain this works then I guess we'll go with it. plan 9 should not be buffering writes. Can you do one more thing, use tcpdump and observe packets. It would be best to have a better understanding of this before we change it.

Reply:
you might also want to run it under strace.

Reply:
I just realized you are using ReadFull now. This is a good fix, thanks for your patience.